### DISCUSSION TITLE
What files should be checked into the repo?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1243] [Q&A] [answered]
There's a general aversion to checking in generated files to version control - you have to deal with additional diffs, source & generated files getting out of sync in the repo, etc. However, without certain files present `npm install` will fail to build. And you can't generate the files needed for it to succeed without first running `npm install` to download `tree-sitter-cli`. Looking around at a few tree-sitter grammar repos it seems most of them check in scanner.c and other files. Is this the recommended way to do it? Or can I modify the `npm install` behavior so that it runs `tree-sitter generate` before building?

**Comment by theHamsta** [comment, marked as answer]
https://github.com/tree-sitter/tree-sitter/issues/930 it's on the 1.0 to remove generated files from the repo. At the moment it's easiest for consumers when C files are on the repo. But this would change with a release mechanism that's identical for all the repos. If you plan improvements for your repo they might be worth to upstream them

**Reply by ahelwer** [reply]
I checked in my generated files and was immediately bit by them being out of sync with the grammar lol. So I added some steps to my CI workflow to check for this; see [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/main/.github/workflows/ci.yml). The relevant portion follows:
```
- name: Generate
  run: node_modules/.bin/tree-sitter generate
- name: Renormalize line endings
  shell: bash
  run: git add --renormalize .
- name: List changes
  shell: bash
  run: git status
  # Command from https://stackoverflow.com/a/62768943/2852699
- name: Check grammar/code sync
  shell: bash
  run: |
    diff_count=$(git status --porcelain=v1 2>/dev/null | wc -l)
    exit $diff_count
```
The step to renormalize line endings is necessary as otherwise the workflow will fail on windows. The "Check grammar/code sync" step basically works by counting the number of files changed then returning that as the bash exit code. If the number of changed files (and by extension the exit code) is nonzero, the workflow will halt with an error.

**Comment by dcreager** [comment]
This has come up again in the TypeScript grammar repo (https://github.com/tree-sitter/tree-sitter-typescript/pull/135, https://github.com/tree-sitter/tree-sitter-typescript/pull/163).

There are a couple of wrinkles there.  One is that that repo actually contains _two_ grammars, and so we had to [update its build](https://github.com/tree-sitter/tree-sitter-typescript/pull/181) scripts to not produce the Rust/NPM bindings files, since `tree-sitter` CLI doesn't know about how we're creating a _single_ Rust/NPM binding containing _both_ of the grammars defined in that repo.

But that also points out that we don't have a single kind of "generated file" that we're dealing with.  The **_generated grammar_** (i.e., `parser.c`, `node-types.json`, etc) is something that primarily depends on the content of the language's `grammar.js` file.  Every now and then there will also be changes that depend on the tree-sitter version you're using, but the proposal in #930 is a good one: no generated files checked into git, produce your derived grammar files using whatever tree-sitter version and language grammar version you need.

The **_binding implementations_** are slightly different.  They're very boilerplate, with a lot of common code that we were copy/pasting across the different language grammar repos.  So the `tree-sitter` CLI grew the ability to produce those for you — but with the intention that you would edit those files as needed, and once edited, the CLI would no longer overwrite them.  You'd be responsible for incorporating in any changes to the "default binding implementation" if those change from tree-sitter version to version.  That's admittedly not as good of a story to start with, but also is something that shouldn't necessarily be the default behavior of the `tree-sitter generate` command.  And it also raises the question of whether the Rust/NPM binding implementations belong in the same repo as the grammars themselves.  (I think they do — the tree-sitter ecosystem already suffers from a [plethora of repos](https://dcreager.net/tree-sitter/map/).)  If the binding implementations live in the repos, but the generated grammars do not, that will add a slight bit of complexity to the build/package/release processes for the bindings.
